flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/rules/logical/LogicalCorrelateToVersionedJoinRule.scala

```diff
@@ -0,0 +1,243 @@
+              leftNode,
+              rightNode,
+              rightVersionedTableCall.versionExpression,
+              rightPrimaryKeyExpression)
+          }
+          else {
+            LogicalVersionedJoin.create(
+              rexBuilder,
+              cluster,
+              logicalCorrelate.getTraitSet,
+              leftNode,
+              rightNode,
+              rightVersionedTableCall.versionExpression,
+              rightVersionExpression,
+              rightPrimaryKeyExpression)
+          })
+        call.transformTo(relBuilder.build())
+      case None =>
+    }
+  }
+
+  private def createRightExpression(
+      rexBuilder: RexBuilder,
+      leftNode: RelNode,
+      rightNode: RelNode,
+      field: String): RexNode = {
+    val rightReferencesOffset = leftNode.getRowType.getFieldCount
+    val rightDataTypeField = rightNode.getRowType.getField(field, false, false)
+    rexBuilder.makeInputRef(
+      rightDataTypeField.getType, rightReferencesOffset + rightDataTypeField.getIndex)
+  }
+}
+
+object LogicalCorrelateToVersionedJoinRule {
+  val INSTANCE: RelOptRule = new LogicalCorrelateToVersionedJoinRule
+}
+
+/**
+  * Simple pojo class for extracted [[TableVersionFunction]] with version expression
+  * extracted by [[VersionExpressionExtractor]].
+  */
+class TableVersionFunctionCall(
+    var tableVersionFunction: TableVersionFunction,
+    var versionExpression: RexNode) {
+}
+
+/**
+  * Find [[TableVersionFunction]] call and run [[VersionExpressionExtractor]] on it's operand.
+  */
+class GetTableVersionFunctionCall(
+    var rexBuilder: RexBuilder,
+    var leftSide: RelNode)
+  extends RexVisitorImpl[TableVersionFunctionCall](false) {
+
+  def visit(node: RexNode): Option[TableVersionFunctionCall] = {
+    val result = node.accept(this)
+    if (result == null) {
+      return None
+    }
+    Some(result)
+  }
+
+  override def visitCall(rexCall: RexCall): TableVersionFunctionCall = {
+    if (!rexCall.getOperator.isInstanceOf[TableSqlFunction]) {
+      return null
+    }
+    val tableFunction = rexCall.getOperator.asInstanceOf[TableSqlFunction]
+    if (!tableFunction.getTableFunction.isInstanceOf[TableVersionFunction]) {
+      return null;
+    }
+    extractVersionExpression(
+      tableFunction.getTableFunction.asInstanceOf[TableVersionFunction],
+      rexCall.getOperands())
+  }
+
+  def extractVersionExpression(
+      tableVersionFunction: TableVersionFunction,
+      operands: util.List[RexNode]): TableVersionFunctionCall = {
+    if (!operands.size().equals(1)) {
+      throw new ValidationException("Table Version Function call must have exactly one argument")
+    }
+    val versionExpressionExtractor =
+      new VersionExpressionExtractor(tableVersionFunction, rexBuilder, leftSide)
+    val versionExpression: RexNode = operands.get(0).accept(versionExpressionExtractor)
+
+    new TableVersionFunctionCall(tableVersionFunction, versionExpression)
+  }
+}
+
+/**
```

Comment:
I'm wondering if we need this extractor. In the end we just need to check if the first operand is a time indicator and get its name, right? Why do we allow CAST operations?

Reply:
`CAST` appears because of Flink's type system.

And yes, we just need to check if the expression is what we are expecting it to be and if so, make sure that there are no weird expressions (like functions calls or other operators) and extract first operand in form of input reference. It was just easier for me to those things with `RexVisitor` compared to manually if checking for tons of conditions.